src/main/java/org/apache/commons/collections4/iterators/ExtendedIterator.java

```diff
@@ -0,0 +1,196 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections4.iterators;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+
+/**
+ * Extends Iterator functionality to include operations commonly found on streams (e.g. filtering, concatenating, mapping).
+ * It also provides convenience methods for common operations.
+ * @param <T> The type of object returned from the iterator.
+ * @since 4.5.0-M3
+ */
+public final class ExtendedIterator<T> implements Iterator<T> {
+    /**
+     * Set to <code>true</code> if this wrapping doesn't permit the use of
+     * {@link #remove()}, otherwise removal is delegated to the base iterator.
+     */
+    private final boolean throwOnRemove;
+
+    /**
+     * Creates an ExtendedIterator wrapped round <code>it</code>,
+     * which does not permit <code>.remove()</code>
+     * even if <code>it</code> does.
+     * @param it The Iterator to wrap.
+     * @return an Extended iterator on {@code it}
+     * @throws UnsupportedOperationException if remove() is called on the resulting iterator.
+     */
+    public static <T> ExtendedIterator<T> createNoRemove(final Iterator<T> it) {
+        return new ExtendedIterator<>(it, true);
+    }
+
+    /**
+     * Creates an ExtendedIterator wrapped round a {@link Stream}.
+     * The extended iterator does not permit <code>.remove()</code>.
+     * <p>
+     * The stream should not be used directly. The effect of doing so is
+     * undefined.
+     * </p>
+     * @param stream the Stream to create an iterator from.
+     * @return an Extended iterator on the {@code stream} iterator.
+     */
+    public static <T> ExtendedIterator<T> create(final Stream<T> stream) {
+        return new ExtendedIterator<T>(stream.iterator(), true);
+    }
+
+    /**
+     * Flattens an iterator of iterators into an Iterator over the next level values.
+     * Similar to list splicing in lisp.
+     * @param iterators An iterator of iterators.
+     * @return An iterator over the logical concatenation of the inner iterators.
+     */
+    public static <T> ExtendedIterator<T> flatten(final Iterator<Iterator<T>> iterators) {
```

Comment:
Missing Javadoc for `T`.

Reply:
Hi @Claudenw 
If you rebase on git master, you can say:
```java
return create(IteratorUtils.chainedIterator(iterators));
```
instead of:
```
        return create(new LazyIteratorChain<T>() {

            @Override
            protected Iterator<? extends T> nextIterator(final int count) {
                return iterators.hasNext() ? iterators.next() : null;
            }

        });
```